site/content/blogs/apprunner-vpc.ja.md

```diff
@@ -0,0 +1,177 @@
+
+AWS Copilotは、様々なタイプの Service をデプロイするために使用できる、いくつかの抽象化された機能を提供します。この記事の例では、[Request-Driven Web Service](https://aws.github.io/copilot-cli/ja/docs/manifest/rd-web-service/) という Copilot パターンを使用しています。これは、受信トラフィックに基づいてオートスケールし、トラフィックがないときはベースラインまでスケールダウンするAWS App Runner Service をデプロイします。このオプションは、リクエスト量が急激に増加する HTTP サービスや、リクエスト量が少ない場合に、より費用対効果が高くなります。
+
+```
+copilot svc init \
+    --app apprunner-vpc  \
+    --svc-type "Request-Driven Web Service" \
+    --name demo-service\
+    --port 5000 \
+    --dockerfile "demo-service/Dockerfile"
+```
+
+通常、AWS Copilot は `manifest.yml` ファイルを生成し、実際のデプロイに進む前にリソースをさらにカスタマイズするために使用することができます。Application で VPC アクセスを有効にするには、Manifest ファイル `copilot/demo-service/manifest.yml` から次のセクションをアンコメントするだけです。
+
+```
+network:
+  vpc:
+    placement: private
+```
+
+これらの追加行は、プライベート VPC リソースに接続するために必要な VPC コネクタを作成するために使用されます。
+
+### VPC 上のデータベース
+
+VPC 内のプライベートリソースに接続できることを確認するために、AWS Copilot を使用して `demoDb` という名前の Amazon Aurora Serverless v1 PostgreSQL データベースクラスタを作成します。
+
+```
+copilot storage init \
+    --name demoDb \
+    --storage-type Aurora\
+    --workload demo-service \
+    --engine PostgreSQL\
+    --initial-db demo
+```
+
+これにより、AWS Copilot CLI を使用してデプロイされる Amazon Aurora Serverless データベースの設定を含む `./copilot/demo-service/addons/demoDb.yml` というファイルが作成されます。
+
+Manifest ファイルを見ると、データベースはデフォルトで Copilot Application と Environment の VPC に関連付けられたプライベートサブネット内に作成されることがわかります。
+
+Copilot は [AWS Secrets Manager](https://aws.amazon.com/jp/secrets-manager/) を使用して、データベースパスワードと Service を生成することに注意してください。データベースパスワードにアクセスできるように、AWS CopilotはシークレットのARNを環境変数にインジェクトし、後で SDK を使用してパスワードを取得できるようにします。今回のデモでは Python を使用しており (なので [boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html) が使用されています)、インジェクトされる環境変数は `DEMO_DB_SECRET_ARN` という名前になります。Python を使ってパスワードを取得するためには、以下のコードを使用します。
+
+```
+sm_client = boto3.client('secretsmanager')
+
+secret_arn = os.getenv("DEMO_DB_SECRET_ARN")
+response = sm_client.get_secret_value(SecretId=secret_arn)
+secret = json.loads(response['SecretString'])
+```
+
+`secret` は以下の情報で作られた Dictionary になります。`{dbClusterIdentifier, password, dbname, engine, port, host, username}` で作られた Dictionary で、これを用いてデータベースに接続することができます。
+
+今回は、SQLAlchemy を利用します。[SQLAlchemy](https://www.sqlalchemy.org/) は Python 用の有名な ORM (Object Relation Mapper) ツール で、低レベルの SQL クエリの代わりにオブジェクト指向パラダイムを使ってデータベースからデータを照会・操作することができます。
+
+データベースに接続するには、[ドキュメントで指定されている](https://flask-sqlalchemy.palletsprojects.com/en/2.x/config/#connection-uri-format)ように、データベース接続文字列を作成する必要があります。この場合、必要なパラメータはすべて `secret` Dictionary から取得することができます。
+
+```
+DB_URI = f"postgresql://{secret['username']}:{secret['password']}@{secret['host']}/{secret['dbname']}"
+```
+
+Copilot は自動的に適切な Instance Role を作成し、作成された Secret に対して `secretsmanager:GetSecretValue` アクションを許可します。また、セキュリティグループも設定する必要はありません。Copilot でデータベースを追加したため、データベースセキュリティグループは、Service セキュリティグループ (VPCコネクタに関連付けられたもの) から PostgreSQL ポート (5432) へのトラフィックを自動的に許可し、Service がアウトバウンドトラフィックをルーティングする場所になります。
```

Comment:
Instance Role -> インスタンスロール because [EC2 documentation](https://docs.aws.amazon.com/ja_jp/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html#roles-usingrole-ec2instance-roles) translates it like that.